level1/cmds/shell_21.asm

```diff
@@ -99,93 +91,113 @@ IcptRtn             stb       <kbdsignl
+* Here's how registers are set when this process is forked:
+*
+*   +-----------------+  <--  Y          (highest address)
+*   !   Parameter     !
+*   !     Area        !
+*   +-----------------+  <-- X, SP
+*   !   Data Area     !
+*   +-----------------+
+*   !   Direct Page   !
+*   +-----------------+  <-- U, DP       (lowest address)
+*
+*   D = parameter area size
+*  PC = module entry point absolute address
+*  CC = F=0, I=0, others undefined
+start               leas      -$05,s          reserve some space on the stack
+                    pshs      y,x,b,a         push registers
+                    ldb       #u0070-1        prepare to clear variables
+                    lbsr      ClearAtU        clear them
+* Install signal handler                    
+                    leax      <IcptRtn,pcr                  point to signal handler
+                    os9       F$Icpt    install it
+                    puls      x,b,a     recover all but Y earlier pushed
+                    std       <parmlen  save the parameter length
+                    beq       noparms@  branch if empty
+                    lbsr      ParseCmdLine     start parsing the command line
+                    bcs       ExitShellWithErr exit if there was an error
                     tst       <u000C
-                    bne       L00C1
-L005B               lds       ,s++
-L005E               leax      <Intro,pcr
-                    tst       <u0012
-                    bne       L0076
-                    bsr       WriteLin
-                    bcs       Exit
-L0069               leax      <DefPrmpt,pcr
-                    ldy       #DefPrmL
-L0070               tst       <u0012
-                    bne       L0076
-                    bsr       WritLin2
-L0076               clra
-                    leax      <u0070,u
-                    ldy       #$00C8
-                    os9       I$ReadLn
-                    bcc       L0094
-                    cmpb      #E$EOF
-                    beq       L00B8
-L0087               tst       <u000F
-                    bne       L008F
-                    tst       <u0014
-                    bne       L00C2
-L008F               os9       F$PErr
-                    bra       L0069
-L0094               cmpy      #$0001
-                    bhi       L00A4
-                    leax      >OS9Prmpt,pcr
-                    ldy       #OS9PrmL
-                    bra       L0070
-L00A4               tst       <u0013
-                    beq       L00AA
-                    bsr       WriteLin
-L00AA               lbsr      L017B
-                    bcc       L0069
-                    tstb
-                    bne       L0087
-                    bra       L0069
+                    bne       ExitShell        exit shell if so
+noparms@            lds       ,s++             get Y off stack pushed earlier (end of parameter area)
+ProcLine            tst       <suppressintro   do we suppress showing the intro?
```

Comment:
@Deek a subtle optimization... I now test the "supressintro" flag before setting X two lines down. The other code set X first and then did the test. It may not justify incrementing the edition byte. Let me know what you think.